File: CfxGenerator/GeneratedFileManager.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Imports System.IO

Public Class GeneratedFileManager

    Private directoryPath As String
    Private oldFiles As New HashSet(Of String)
    Private newFiles As New List(Of String)

    Private licenseStub As String

    Public Sub New(directoryPath As String)
        Me.directoryPath = directoryPath
        If Not Directory.Exists(directoryPath) Then
            Directory.CreateDirectory(directoryPath)
        End If
        Dim files = Directory.GetFiles(directoryPath)
        For Each f In files
            oldFiles.Add(f)
        Next

        licenseStub = AssemblyResources.GetString("LicenseStub")

    End Sub

    Public Sub New()
        'need a dummp writer
        directoryPath = Nothing
    End Sub

    Public Sub WriteFileIfContentChanged(filename As String, content As String)

        If directoryPath = Nothing Then Return

        Dim filepath = Path.Combine(directoryPath, filename)

        If newFiles.Contains(filepath) Then
            Stop
        End If

        Dim b = New CodeBuilder
        b.AppendLine(licenseStub)
        b.AppendComment("Generated file. Do not edit.")
        b.AppendLine()
        b.AppendLine()
        b.Append(content)

        content = b.ToString()

        Dim oldcontent = String.Empty
        If oldFiles.Contains(filepath) Then
            oldcontent = File.ReadAllText(filepath)
            oldFiles.Remove(filepath)
        End If

        If Not content.Equals(oldcontent) Then
            File.WriteAllText(filepath, content)
        End If

        newFiles.Add(filepath)

    End Sub

    Public Sub DeleteObsoleteFiles()
        For Each f In oldFiles
            File.Delete(f)
        Next
    End Sub

    Public Function GetNewFiles() As String()
        Return newFiles.ToArray()
    End Function



    Shared Sub PatchFilesLicense(dir As String)

        Dim licenseStub = AssemblyResources.GetString("LicenseStub")
        Dim licenseVB = licenseStub.Replace("//", "''")

        Dim files = Directory.GetFiles(dir)
        For Each f In files
            Dim ext = Path.GetExtension(f)
            If New String() {".cs", ".vb", ".c", ".cpp", ".h"}.Contains(ext.ToLowerInvariant()) Then
                Dim content = File.ReadAllText(f)

                If Not content.StartsWith("'' Copyright") AndAlso Not content.StartsWith("// Copyright") Then
                    Dim b = New CodeBuilder
                    If ext = ".vb" Then
                        b.AppendLine(licenseVB)
                    Else

                        b.AppendLine(licenseStub)
                    End If

                    b.AppendLine()
                    b.AppendLine()
                    b.Append(content)

                    content = b.ToString()

                    Debug.Print(f)
                    File.WriteAllText(f, content)

                Else
                    Debug.Assert(content.StartsWith(licenseVB) OrElse content.StartsWith(licenseStub))
                End If
            End If
        Next

        Dim dirs = Directory.GetDirectories(dir)
        For Each d In dirs
            If Not d.EndsWith(".hg") AndAlso Not d.Contains("\Generated\") AndAlso Not d.Contains("\cef\") Then
                PatchFilesLicense(d)
            End If
        Next


    End Sub


End Class

File: CfxGenerator/GeneratedFileManager.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

public class GeneratedFileManager {
    private string directoryPath;
    private HashSet<string> oldFiles = new HashSet<string>();

    private List<string> newFiles = new List<string>();

    private string licenseStub;

    public GeneratedFileManager(string directoryPath) {
        this.directoryPath = directoryPath;
        if(!Directory.Exists(directoryPath)) {
            Directory.CreateDirectory(directoryPath);
        }
        var files = Directory.GetFiles(directoryPath);
        foreach(var f in files) {
            oldFiles.Add(f);
        }

        licenseStub = AssemblyResources.GetString("LicenseStub");
    }

    public GeneratedFileManager() {
        //need a dummp writer
        directoryPath = null;
    }

    public void WriteFileIfContentChanged(string filename, string content) {
        if(directoryPath == null)
            return;

        var filepath = Path.Combine(directoryPath, filename);

        if(newFiles.Contains(filepath)) {
            System.Diagnostics.Debugger.Break();
        }

        var b = new CodeBuilder();
        b.AppendLine(licenseStub);
        b.AppendComment("Generated file. Do not edit.");
        b.AppendLine();
        b.AppendLine();
        b.Append(content);

        content = b.ToString();

        var oldcontent = string.Empty;
        if(oldFiles.Contains(filepath)) {
            oldcontent = File.ReadAllText(filepath);
            oldFiles.Remove(filepath);
        }

        if(!content.Equals(oldcontent)) {
            File.WriteAllText(filepath, content);
        }

        newFiles.Add(filepath);
    }

    public void DeleteObsoleteFiles() {
        foreach(var f in oldFiles) {
            File.Delete(f);
        }
    }

    public string[] GetNewFiles() {
        return newFiles.ToArray();
    }

    public static void PatchFilesLicense(string dir) {
		
        var licenseStub = AssemblyResources.GetString("LicenseStub");

        var files = Directory.GetFiles(dir);
        foreach(var f in files) {
            var ext = Path.GetExtension(f);
            if(new string[] {
				".cs",
				".c",
				".cpp",
				".h"
			}.Contains(ext.ToLowerInvariant())) {
                var content = File.ReadAllText(f);

                if(true || !content.StartsWith(licenseStub)) {

                    if(content.StartsWith("// Copyright")) {
                        var ex = new Regex(@"^// Copyright.*?txt file for details.\s*", RegexOptions.Singleline);
                        Debug.Assert(ex.IsMatch(content));
                        content = ex.Replace(content, licenseStub + Environment.NewLine);
                    } else {

                        var b = new CodeBuilder();
                        b.AppendLine(licenseStub);
                        b.Append(content);

                        content = b.ToString();

                    }


                    Debug.Print(f);
                    File.WriteAllText(f, content);
                }
            }
        }

        var dirs = Directory.GetDirectories(dir);
        foreach(var d in dirs) {
            if(!d.EndsWith(".hg") && !d.Contains("\\Generated\\") && !d.Contains("\\cef\\")) {
                PatchFilesLicense(d);
            }
        }
    }
}